polytools: Added subresultant polynomials and coefficients

mmaaz-git opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
#26785


#### Brief description of what is fixed or changed

Added functions for subresultant polynomials and subresultant coefficients. These use the subresultant PRS (`subresultants()`) already in SymPy. The subresultant polynomials are computed from the subresultant PRS according to a certain rule, described in the comments. As for the subresultant coefficients, we simply take the coefficient of $x^{i-1}$ of the $i$-th subresultant polynomial. These functions are necessary for other tasks, including eventually for cylindrical algebraic decomposition.

For reference, the three subresultant functions line up with the following Mathematica commands:
- `subresultants` -> `SubresultantPolynomialRemainders`
- `subresultant_polys` -> `SubresultantPolynomials`
- `subresultant_coeffs` -> `Subresultants`


#### Technical details

Given the subresultant PRS, the remainder $r_i$ is the $deg(r_{i-1})$-th subresultant polynomial. Additionally, if $deg(r_i) < deg(r_{i-1}) - 1$, then the $deg(r_i)$ subresultant polynomial is $r_i * LC(r_i) ^ {c_i}$, where $c_i = deg(r_{i-1})-deg(r_i)-1$.

Given the subresultant polynomials, the coefficient of $x^{i-1}$ in the $i$-th subresultant polynomial is the $i$-th subresultant coefficient.

#### Other comments

Currently, `subresultants()` is in the main sympy namespace, i.e. it is imported when `from sympy import *`. Should these two be added to that as well? 


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* polys
  * Added subresultant polynomials and subresultant coefficients.
<!-- END RELEASE NOTES -->

mmaaz-git committed:
polytools: Added subresultant polynomials and coefficients

Added functions for subresultant polynomials and subresultant
coefficients. These use the subresultant PRS (`subresultants()`) already
in SymPy. The subresultant polynomials are computed from the subresultant
PRS according to a certain rule, described in the comments. As for the
subresultant coefficients, we simply take the coefficient of x^{i-1} of
the ith subresultant polynomial. These functions are necessary for other
tasks, including eventually for cylindrical algebraic decomposition.

mmaaz-git committed:
Typo fix in comments

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2669,124 @@ def subresultants(f, g):
 
         return list(map(per, result))
 
+
+    def subresultant_polys(f, g):
```
It should be one blank line between methods like the methods above.

mmaaz-git replied:
Done

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2669,124 @@ def subresultants(f, g):
+        >>> Poly(x**2 + 1, x).resultant(Poly(x**2 - 1, x))
+        4
+
+        """
```
It would be good to add a See Also section here that references related functions.

oscarbenjamin commented:
> Currently, `subresultants()` is in the main sympy namespace, i.e. it is imported when `from sympy import *`. Should these two be added to that as well?

Probably.

What I'm not sure about is the names of these functions. Does anyone have any views on what would be good names for them?

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2669,124 @@ def subresultants(f, g):
+        if len(prs) <= 1:
+            return []
+
+        subres_polys = [0] * (m + 1)
+
+        for i in reversed(range(2, len(prs))):
+            # remainder r_i is the deg(r_{i-1})-th subres poly
+            subres_polys[ prs[i-1].degree() -1 ] = prs[i]
+
+            # if there is a "degree jump"
+            # if deg(r_i) < deg(r_{i-1}) - 1, then the deg(r_i) subres poly
+            # is r_i * LC(r_i) ^ c_i, where c_i = deg(r_{i-1})-deg(r_i)-1
+            if prs[i].degree() < prs[i-1].degree() - 1:
+                degree_jump = prs[i-1].degree() - prs[i].degree() - 1
+                subres_polys[ prs[i].degree() ]  =\
+                    prs[i] * prs[i].LC() ** degree_jump
+
+        # handle last one
+        subres_polys[-1] = prs[1] * g.LC() ** (n - m - 1)
+
+        # want it to always return Poly objects for consistency
+        # this basically just makes sure 0 is returned as Poly(0)
+        for i in range(len(subres_polys)):
+            if not isinstance(subres_polys[i], Poly):
+                subres_polys[i] = Poly(i, f.gen)
```
I think this code could be simplified by using a single loop that appends items one at a time.

mmaaz-git committed:
Fixed line spacing

mmaaz-git committed:
Added see also references

mmaaz-git commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2669,124 @@ def subresultants(f, g):
+        # handle last one
+        subres_polys[-1] = prs[1] * g.LC() ** (n - m - 1)
+
+        # want it to always return Poly objects for consistency
+        # this basically just makes sure 0 is returned as Poly(0)
+        for i in range(len(subres_polys)):
+            if not isinstance(subres_polys[i], Poly):
+                subres_polys[i] = Poly(i, f.gen)
```
Ah you're right, the issue was in the initialization of the list which was previously `[0 for _ in range(m + 1)]`, but it should actually be `[Poly(0, f.gen) for _ in range(m + 1)]`. Will fix.

mmaaz-git committed:
Initialize subres polys as Poly(0)

mmaaz-git committed:
Refactored subresultant_polys loop

mmaaz-git committed:
Fixed a doctest (checked by hand)

mmaaz-git committed:
Added see also in the public functions

mmaaz-git commented:
I've simplified and refactored the loop. For example there's no need to traverse backwards, many quantities are reused, 
 so I precompute at the beginning of the loop, and so on. 

However, if I understand your comment correctly: it might be clunky to make the loop append one item at a time, as opposed to initializing a list of zeros. One way to make it append one at a time is that when you compute a polynomial, store the corresponding degree (its place in the list basically), and then at the end, put them all together sorted by degree, filling in with zero if that degree polynomial is missing. But I'm not sure if this is better?

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2675,128 @@ def subresultants(f, g):
+            # if there is a "degree jump"
+            # if deg(r_i) < deg(r_{i-1}) - 1, then the deg(r_i) subres poly
+            # is r_i * LC(r_i) ^ c_i, where c_i = deg(r_{i-1})-deg(r_i)-1
+            degree_jump = prev_deg - curr_deg - 1
+            if degree_jump > 0:
+                subres_polys[curr_deg] = curr * curr.LC() ** degree_jump
```
Are there any test cases in which the degree jump is greater than 1?

There don't seem to be any and I am trying to work out exactly what the behaviour should be in that case.

mmaaz-git replied:
Here's an example:
```
In [16]: subresultants(x**5 + x**3 - 1, x**3)
Out[16]: [x**5 + x**3 - 1, x**3, 1]
```

At `i=2`, we have prev_deg-1 = 2, so the 2nd element of subres_polys is x^3. Furthermore, the degree jump is 3-0-1=2. So, the 0th element of subres_polys is 1 * 1 ^ 2 = 1. 

So that:
```
In [19]: subresultant_polys(x**5 + x**3 - 1, x**3)
Out[19]: [1, 0, 1, x**3]
```

In this case it seems inconsequential because its just 1. But it becomes important in the multivariate setting.
```
In [18]: subresultant_polys(x**5 + x**3 - 1, y*x**3)
Out[18]: [y**3, 0, y**3, x**3*y]
```
Where the 0th element is now y * y^2 = y^3.

mmaaz-git replied:
Here's an example:
```
In [19]: subresultants(x**5 + x**3 - 1, 2*x**3)
Out[19]: [x**5 + x**3 - 1, 2*x**3, 8]
```

At i=2, the degree jump deg(r_{i-1}) - deg(r_i) - 1 = 2. I've now added this as a test case. But, in this case, the 0th element was not matching the Mathematica output for the same (it was 512 I think, vs 32 which is correct, so basically off by some exponents). I spent a lot of time digging around on what the right form of this formula is, but I couldn't figure it out. I'm sure I'm missing something minor, but I'm not sure what. But eventually, I just used the fact that the 0th element is always the resultant. It's now giving correct output but I'm not sure if this is fully right.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2675,128 @@ def subresultants(f, g):
+        if len(prs) <= 1:
+            return []
+
+        subres_polys = [Poly(0, f.gen) for _ in range(m + 1)]
+
+        subres_polys[-1] = g * g.LC() ** (n - m - 1)
+
+        for i in range(2, len(prs)):
+            curr = prs[i]
+            prev = prs[i-1]
+            curr_deg = curr.degree()
+            prev_deg = prev.degree()
+
+            # remainder r_i is the deg(r_{i-1})-th subres poly
+            subres_polys[ prev_deg -1 ] = curr
+
+            # if there is a "degree jump"
+            # if deg(r_i) < deg(r_{i-1}) - 1, then the deg(r_i) subres poly
+            # is r_i * LC(r_i) ^ c_i, where c_i = deg(r_{i-1})-deg(r_i)-1
+            degree_jump = prev_deg - curr_deg - 1
+            if degree_jump > 0:
+                subres_polys[curr_deg] = curr * curr.LC() ** degree_jump
```
I was thinking like:
```python
zero = f.zero
subres_polys = []
deg = prs[-1].degree()

for p in reversed(prs[1:-1]):
    deg_new = p.degree()
    jump = deg_new - deg - 1
    if jump > 0:
        subres_polys.extend([zero] * jump)
        subres_polys[deg_new-1] = deg_new * p.LC() * jump**deg_new
    subres_polys.append(p)
    deg = deg_new
```
I'm sure this isn't quite right but that is sort of how I imagined the code would look.

mmaaz-git replied:
Thanks for the suggestion! I tried to incorporate the spirit of this. Basically, the list has to be long enough to assign the $deg(r_{i-1})$-th element. So, what I do now is append zeros until it is one short of that length, then append the required element (the remainder r_i). Then, the degree jump case is handled after in the loop, because it is guaranteed that the required index has already been created in the list.

mmaaz-git committed:
polytools: Refactored subresultant polynomial code

Previously, subresultant_polys() initialized a list of 0s and then
assigned appropriate elements. This has now been changed to appending
or extending the list as needed. This happens by keeping track of the
degrees of the subresultant polys that are being computed versus the
current length of the list, and basically appends 0s until the
appropriate length. Also, I fixed an issue that was appearing with the
0th element of the subresultant polys, where it was being scaled by
an incorrect factor. It should be correct now, due to the observation
that the 0th element is always just the resultant. Also, there is no
need to store r_{i-1} (previously called "prev"), as only its degree
is needed.

mmaaz-git committed:
Added new test case with degree jump

mmaaz-git committed:
Added resultants to see also for subresultants

mmaaz-git committed:
Changed Poly(0) to f.zero

mmaaz-git committed:
Minor fix for code quality

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2676,140 @@ def subresultants(f, g):
+        This uses the subresultant PRS. The remainder r_i is the deg(r_{i-1})-th
+        subresultant polynomial. Additionally, if deg(r_i) < deg(r_{i-1}) - 1,
+        then the deg(r_i) subresultant polynomial is r_i * LC(r_i) ^ c_i, where
+        c_i = deg(r_{i-1})-deg(r_i)-1.
```
Things like `r_i * LC(r_i) ^ c_i` either need to be rendered as code (double backticks) or LaTeX (single backticks or dollars).

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -2669,6 +2676,140 @@ def subresultants(f, g):
 
         return list(map(per, result))
 
+    def subresultant_polys(f, g):
+        """
+        Computes the subresultant polynomials of two polynomials ``f`` and ``g``.
+        The returned list has length min(deg(f), deg(g))+1, and is in order of
+        increasing degree. The first polynomial in the list is the resultant.
```
The first line of the docstring should be a single line sentence as the one line summary

https://docs.sympy.org/latest/contributing/docstring.html#single-sentence-summary

I would say here that it should be:
```
Subresultant polynomials of ``f`` and ``g`` with respect to the main variable.
```

oscarbenjamin commented:
I think this looks good now apart from formatting the docstring.

oscarbenjamin commented:
> I think this looks good now apart from formatting the docstring.

I'm marking this as draft for now. When it is ready for another review mark it as ready for review and leave a comment.

oscarbenjamin commented:
It would be great to get this finished. I can't remember where we were with this because there were a few different pull requests...
